### DISCUSSION TITLE
Why this `cannot borrow multipart as mutable more than once at a time` error?

### DISCUSSION THREAD

**Discussion by frederikhors** [tokio-rs/axum#1644] [Q&A]
I have this code:

```rust
pub async fn upload(
    mut multipart: Multipart,
) -> Result<Response, StatusCode> {
    while let Some(field) = multipart.next_field().await.unwrap() {
        let filename = if let Some(filename) = field.file_name() {
            filename.to_string()
        } else {
            continue;
        };

        let body_with_io_error = field.map_err(|err| io::Error::new(io::ErrorKind::Other, err));

        let body_reader = StreamReader::new(body_with_io_error);

        // futures::pin_mut!(body_reader);

        let body_reader = Box::pin(body_reader);

        let res = client.upload(filename, body_reader).await?;

        return Ok(Response::builder().status(StatusCode::CREATED).unwrap());
    }

    Err(StatusCode::INTERNAL_SERVER_ERROR)
}
```

but the `client.upload()` signature is:

```rust
pub async fn upload(filename: &str, reader: Pin<Box<(dyn AsyncRead + Send + Sync)>>) -> Result<(), ()>;
```

It error with:

```log
error[E0499]: cannot borrow `multipart` as mutable more than once at a time
   |
29 |     while let Some(field) = multipart.next_field().await.unwrap() {
   |                             ^^^^^^^^^^^^^^^^^^^^^^ `multipart` was mutably borrowed here in the previous iteration of the loop
...
60 |                 body_reader,
   |                 ----------- cast requires that `multipart` is borrowed for `'static`
```

How can I fix this?

**Comment by davidpdrsn** [comment]
The signature of `next_field` is

```
pub async fn next_field(&mut self) -> Result<Option<Field<'_>>, MultipartError>
```

The `<'_>` in `Field<'_>` means it contains a reference to the `&mut self` in this case. Think of it as `Field<'multipart>` where `'multipart` is the lifetime of `multipart`.

We do this because you can only consume one field at a time. `Field` is sorta represents a "pointer" into the streaming request body and you cannot consume the stream from multiple places at once. axum enforces this statically whereas `multer` (the library axum uses under the hood) checks it at runtime https://docs.rs/multer/latest/multer/struct.Multipart.html#field-exclusivity.

Next `Box<dyn T>` has an implicit `'static` so it is really `Box<dyn T + 'static>`. That means `T` cannot contain references, so it can't be like `Vec<&'a Foo>` or `Field<'multipart>`. So `Pin<Box<(dyn AsyncRead + Send + Sync)>>` is `Pin<Box<(dyn AsyncRead + Send + Sync + 'static)>>`. That doesn't work because `Field<'_>` isn't `'static` because it contains a reference to `multipart`.

I'm not sure why the library uses that signature, could just be an oversight or maybe there is some technical reason for it. I'm honestly not sure if there is a way to resolve it except for buffering the field bytes in memory with https://docs.rs/axum/latest/axum/extract/multipart/struct.Field.html#method.bytes but I'm assuming you want to avoid that.

**Reply by davidpdrsn** [reply]
For testing I tried changing the code to

```rust
pub async fn upload(mut multipart: Multipart) -> Result<Response, StatusCode> {
    while let Some(field) = multipart.next_field().await.unwrap() {
        let filename = if let Some(filename) = field.file_name() {
            filename.to_string()
        } else {
            continue;
        };

        let body_with_io_error = field.map_err(|err| io::Error::new(io::ErrorKind::Other, err));

        let body_reader = StreamReader::new(body_with_io_error);

        // futures::pin_mut!(body_reader);

        let body_reader = Box::pin(body_reader);

        let res = Client.upload(&filename, body_reader).await;

        // return Ok(Response::builder().status(StatusCode::CREATED).unwrap());
    }

    Err(StatusCode::INTERNAL_SERVER_ERROR)
}

struct Client;

impl Client {
    async fn upload<'a>(
        &self,
        filename: &str,
        reader: Pin<Box<(dyn AsyncRead + Send + Sync + 'a)>>,
                                                    // ^^ adding this
    ) -> Result<(), ()> {
        todo!()
    }
}
```

which does compile

**Reply by frederikhors** [reply]
Thank you!

The `'a` does work, but now I have an issue in one adapter with: `.body(reqwest::Body::wrap_stream(ReaderStream::new(reader)))` (using reqwest) which errors with:

```log
error[E0521]: borrowed data escapes outside of associated function
   |
47 |     async fn put_file<'a>(
   |                       -- lifetime `'a` defined here
48 |         &'a self,
   |         -------- `self` is a reference that is only valid in the associated function body
...
88 |             .body(reqwest::Body::wrap_stream(ReaderStream::new(reader)))
   |                   ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |                   |
   |                   `self` escapes the associated function body here
   |                   argument requires that `'a` must outlive `'static`
```

**Reply by davidpdrsn** [reply]
Wait, where

```rust
pub async fn upload(filename: &str, reader: Pin<Box<(dyn AsyncRead + Send + Sync)>>) -> Result<(), ()>;
```

even come from? If is search for `fn upload` in https://github.com/durch/rust-s3 nothing comes up.

**Reply by frederikhors** [reply]
Nope, this is one of mine. I dropped rust-s3, I'm using reqwest directly using the insutrctions here (https://github.com/paolobarbolini/rusty-s3/discussions/66#discussioncomment-4403888).

**Reply by davidpdrsn** [reply]
Thats the same issue. `reqwest` requires a stream that is fully owned (i.e. is `'static`, i.e. doesn't contain references). Not sure how to fix it in that case.

**Reply by frederikhors** [reply]
Apart from this last (I hope) error I'm thinking about changing my signature (`AsyncRead` was there because of initial try with axum multipart field).

But since I'm using two adapters, one for FS:

```rust

let path = Path::new(&self.files_dir).join(filename);

Ok(Box::pin(tokio::fs::File::open(path).await?))

```

and the one for S3, I don't know which signature to use.

I thought:

```rust
reader: Pin<Box<dyn Stream<Item = Result<Bytes>> + Send + Sync>>,
```

what do you think?

**Reply by frederikhors** [reply]
Can I remove `AsyncRead` at all?

**Reply by davidpdrsn** [reply]
You can't use `Stream` because `File` doesn't implement that.

You haven't provided much context for that so can't say more.

**Reply by frederikhors** [reply]
Ok, I'll remove the File adapter. How is the signature for Stream instead of AsyncRead?

**Reply by davidpdrsn** [reply]
I can't really much without more context. As a signature on its own it looks fine I guess 🤷

**Reply by frederikhors** [reply]
To give you more context that you rightfully need I've created [a small Github repo](https://github.com/frederikhors/axum-upload-local-and-cloud) that mimics exactly what I'm trying to achieve: streaming file uploads to both file systems and the cloud (S3) using `axum` (and `reqwest`).

I'll PR axum's examples as soon I can understand how to complete this small project.

I can change all the code you see. There is no restriction. All signatures were invented by me who are in my early days with the awesome Rust.

If you want I can also eliminate `reqwest`, as long as I understand what to use instead! :smile:

UPDATE:

I tried with `hyper` too and it's the same error obviously:

```rust
hyper::Client::new()
    .request(
        hyper::Request::builder()
            .uri(action.sign(presigned_url_duration).to_string())
            .body(hyper::Body::wrap_stream(ReaderStream::new(reader)))
            .unwrap(),
    )
    .await
    .unwrap();
```

```log
error[E0521]: borrowed data escapes outside of associated function
    |
66  |     async fn put_file<'a>(
    |                       -- lifetime `'a` defined here
67  |         &'a self,
    |         -------- `self` is a reference that is only valid in the associated function body
...
124 |                     .body(hyper::Body::wrap_stream(ReaderStream::new(reader)))
    |                           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |                           |
    |                           `self` escapes the associated function body here
    |                           argument requires that `'a` must outlive `'static`
```

**Reply by davidpdrsn** [reply]
Yes because that method has an `S: Stream<Item = Result<O, E>> + Send + 'static,` bound.

Why are you even using multipart for `upload` if you're only handling the first field by the way? If your only handling one "part" then couldn't that be the whole request body? Like you're using a `while` loop but always returning the first field.

**Reply by frederikhors** [reply]
The client they are using is sending a multipart request. What are you suggesting? To avoid multipart and use https://docs.rs/axum/0.6.1/axum/extract/struct.BodyStream.html instead?

**Reply by davidpdrsn** [reply]
Yes. If you're only sending one piece of data then using multipart isn't necessary, as thats designed for sending multiple things.

**Reply by frederikhors** [reply]
@davidpdrsn I'm trying avoiding multipart and it works amazingly well! **I'm happy!**

Just one thing: how can I get the filename now? The filename before was in multipart field "name".

I was using this code:

```rust
while let Some(field) = multipart.next_field().await.unwrap() {
  let filename = if let Some(filename) = field.file_name() {
      filename.to_string()
  } else {
      continue;
  };
// ...
}
```

Now I'm using this code:

```rust
pub async fn upload(
    stream: BodyStream,
    TypedHeader(content_length): TypedHeader<ContentLength>,
) -> Result<Response, StatusCode> {
    let stream = stream.map_err(|err| io::Error::new(io::ErrorKind::Other, err));

    let created_file = file_create.execute(
            "filename".to_string(),
            content_length.0,
            Box::pin(StreamReader::new(stream)),
        )
        .await?;

    Ok(Response::builder().status(StatusCode::CREATED).unwrap())
}
```

What to use now?

**Reply by davidpdrsn** [reply]
You could send that in a query parameter instead maybe.

**Reply by frederikhors** [reply]
The thing is they are using a javascript client that sends multipart form. I cannot decide for that side... Is there a way to detect the filename?

**Reply by davidpdrsn** [reply]
> The thing is they are using a javascript client that sends multipart form

Well in that case you can't just switch to `BodyStream`. I didn't realize you didn't control the client.

**Reply by praveenperera** [reply]
@frederikhors @davidpdrsn  was there a solution to this, i'm also trying to stream a field from multipart to a reqwest body for upload.

```rust
    // find the field and stream upload to GCS
    while let Some(field) = multipart.next_field().await.wrap_err("cant get field")? {
        let name = field.name().unwrap_or_default().to_string();

        if name != "file" {
            continue;
        }

        let body = Body::wrap_stream(field);

        client
            .upload_multiple_chunk(body, &chunk_size)
            .await
            .wrap_err("error uploading chunk")?;

        return Ok((StatusCode::OK, "ok"));
    }
```

I get the same error on multipart needing to be static, upload_multiple_chunk expects `Into<Body>`
